src/persist-client/src/machine.rs

```diff
@@ -0,0 +1,400 @@
+                if &write_cap.upper != desc.lower() {
+                    return Err(InvalidUsage(anyhow!("WIP")));
+                }
+                write_cap.upper.clone_from(desc.upper());
+                // WIP have to trim desc
+                state.trace.push((key.to_owned(), desc.clone()));
+                Ok(())
+            })
+            .await?;
+        if let Err(err) = res {
+            return Ok(Err(err));
+        }
+        Ok(Ok(seqno))
+    }
+
+    pub async fn downgrade_since(
+        &mut self,
+        deadline: Instant,
+        reader_id: &ReaderId,
+        new_since: &Antichain<T>,
+    ) -> Result<Result<SeqNo, InvalidUsage>, LocationError> {
+        let (seqno, res) = self
+            .apply_unbatched_cmd(deadline, |_, state| {
+                let read_cap = state
+                    .readers
+                    .get_mut(reader_id)
+                    // WIP this is more likely that the lease expired
+                    .ok_or_else(|| InvalidUsage(anyhow!("reader not registered")))?;
+                if !PartialOrder::less_equal(&read_cap.since, new_since) {
+                    return Err(InvalidUsage(anyhow!("WIP")));
+                }
+                read_cap.since.clone_from(new_since);
+                Ok(())
+            })
+            .await?;
+        if let Err(err) = res {
+            return Ok(Err(err));
+        }
+        self.maybe_compact_log(deadline).await?;
+        Ok(Ok(seqno))
+    }
+
+    pub async fn deregister_writer(
+        &mut self,
+        deadline: Instant,
+        writer_id: &WriterId,
+    ) -> Result<SeqNo, LocationError> {
+        let (seqno, ()) = self
+            .apply_unbatched_cmd(deadline, |_, state| {
+                state.writers.remove(writer_id);
+            })
+            .await?;
+        Ok(seqno)
+    }
+
+    pub async fn deregister_reader(
+        &mut self,
+        deadline: Instant,
+        reader_id: &ReaderId,
+    ) -> Result<SeqNo, LocationError> {
+        let (seqno, ()) = self
+            .apply_unbatched_cmd(deadline, |_, state| {
+                state.readers.remove(reader_id);
+            })
+            .await?;
+        self.maybe_compact_log(deadline).await?;
+        Ok(seqno)
+    }
+
+    pub async fn handle_compact_trace_res(
+        &mut self,
+        deadline: Instant,
+        res: CompactTraceRes<T>,
+    ) -> Result<SeqNo, LocationError> {
+        let (seqno, ()) = self
+            .apply_unbatched_cmd(deadline, |_, state| todo!("{:?} {:?}", state, res))
+            .await?;
+        Ok(seqno)
+    }
+
+    async fn apply_unbatched_cmd<R, WorkFn: FnMut(SeqNo, &mut State<T>) -> R>(
+        &mut self,
+        deadline: Instant,
+        mut work_fn: WorkFn,
+    ) -> Result<(SeqNo, R), LocationError> {
+        let (_, _, seqno, ret) = self
+            .collection
+            .update_metadata(deadline, |durable_seqno, durable_state| {
+                if self.seqno != durable_seqno {
```

Comment:
sanity checking understanding here: the invariant should be `durable_seqno >= self.seqno` and then if it happens to be greater than we have to reset our state right?

If the durable_seqno is somehow less than self.seqno then something is very wrong right?

Reply:
yeah that sounds right